urlcaef331f4e_dreamable_cra_tgz-pJ8-coho_lp_project_LPProjectJ8/src/coho/common/number/BasicType.java

```diff
@@ -0,0 +1,99 @@
+package coho.common.number;
+
+/*
+ * If you want to add a subclass of CohoNumber, change this class
+ */
+public abstract class BasicType implements CohoType {
+	protected Class classType;
+	protected CohoNumber one;
+	protected CohoNumber zero;
+	protected BasicType(Class classType, CohoNumber zero, CohoNumber one){
+		this.classType = classType;
+		this.zero = zero;
+		this.one = one;
+	}
+	public Class classType(){
+		return classType;
+	}
+	public String name(){
+		return classType.getName();
+	}
+	public CohoNumber zero(){
+		return zero;
+	}
+	public CohoNumber one(){
+		return one;
+	}
+	
+	private static int typeKey(CohoType t){
+		if(t instanceof ScaleType)
+			return 0;
+		if(t instanceof IntervalType)
+			return 1;
+		throw new UnsupportedOperationException("We only support Scale Number and Interval Number now");
+	}
+
+	public BasicType promote(CohoType that) {
+		int p1 = typeKey(this);
+		int p2 = typeKey(that);
+		if(p1==0&&p2==0)
+			return ((ScaleType)this).promote((ScaleType)that);//this will not be called
+		else if(p1==1&&p2==1)
+			return ((IntervalType)this).promote((IntervalType)that);//this will not be called
+		else if(p1==1)
+			return this.promote(IntervalType.promote((ScaleType)that));
+		else// if(p2==1)
+			return IntervalType.promote((ScaleType)this).promote(that);
+	}
+	
+	public static CohoNumber[] promote(CohoNumber a, CohoNumber b){
+		int p1 = typeKey(a.type());
+		int p2 = typeKey(b.type());
+		if(p1==0&&p2==0)
+			return ScaleType.promote((ScaleNumber)a, (ScaleNumber)b);
+		else if(p1==1&&p2==1)
+			return IntervalType.promote((IntervalNumber)a,(IntervalNumber)b);
+		else if(p1==1)
+			return IntervalType.promote((IntervalNumber)a, IntervalType.promote((ScaleNumber)b));
+		else
+			return IntervalType.promote(IntervalType.promote((ScaleNumber)a),(IntervalNumber)b);
+			
+	}
+	public static CohoNumber[] promote(CohoNumber a, Number b){
+		int p = typeKey(a.type());
+		if(p==0)
+			return ScaleType.promote((ScaleNumber)a, b);
+		else 
+			return IntervalType.promote((IntervalNumber)a, b);
+	}
+	
+	public static CohoNumber promoteOp(CohoNumber n1, CohoNumber n2, CohoNumber.ArithOp op){
+		CohoNumber[] operands = promote(n1,n2);
+		int t = typeKey(operands[0].type());
+		if(t==0){
+			ScaleNumber op1 = (ScaleNumber)operands[0];
+			ScaleNumber op2 = (ScaleNumber)operands[1];
+			return ScaleType.promoteOp(op1,op2,op);
+		}else{
+			IntervalNumber op1 = (IntervalNumber)operands[0];
+			IntervalNumber op2 = (IntervalNumber)operands[1];
+			return IntervalType.promoteOp(op1,op2,op);			
+		}		
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 71:81, 84:94.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.